File: src/com/ds/tree/BinarySearchTree.java
package com.ds.tree;

import java.util.ArrayList;
import java.util.List;

public class BinarySearchTree {
  
  public Node root;
  
  public List<Integer> list = new ArrayList<>();
  
  public BinarySearchTree() {
    root = null;
  }
  
  private int size(Node node) {
    if (node == null) return 0;
    return node.count;
  }
  
  private Node put(Node node, int value) {
    if (node == null) {
      return new Node(value);
    }
  
    if (value < node.value) {
      node.left = put(node.left, value);
      node.left.parent = node;
    } else if (value > node.value) {
      node.right = put(node.right, value);
      node.right.parent = node;
    } else {
      node.value = value;
    }
    node.count = size(node.left) + size(node.right) + 1;
    return node;
  }
  
  private Node deleteMin(Node node) {
    if (node.left == null) return node.right;
    node.left = deleteMin(node.left);
    node.count = 1 + size(node.left) + size(node.right);
    return node;
  }
  
  private Node deleteMax(Node node) {
    if (node.right == null) return node.left;
    node.right = deleteMax(node.right);
    node.count = 1 + size(node.right) + size(node.left);
    return node;
  }
  
  private Node delete(Node node, int value) {
    if (node == null) return null;
    
    if (value < node.value) {
      node.left = delete(node.left, value);
    } else if (value > node.value) {
      node.right = delete(node.right, value);
    } else {
      if (node.left == null) {
        return node.right;
      } else if (node.right == null) {
        return node.left;
      }
  
      node.value = getMin(node.right);
      node.right = delete(node.right, node.value);
    }
    node.count = size(node.left) + size(node.right) + 1;
    return node;
  }
  
  public void put(int value) {
    root = put(root, value);
  }
  
  public void delete(int value) {
    root = delete(root, value);
  }
  
  public void deleteMin() {
    root = deleteMin(root);
  }
  
  public void deleteMax() {
    root = deleteMax(root);
  }
  
  public Node get(Node root, int value) {
    Node node = root;
    if (node.value == value) {
      return node;
    } else if (value < node.value) {
      node = node.left;
      get(node, value);
    } else {
      node = node.right;
      get(node, value);
    }
    return node;
  }
  
  public int getMin(Node root) {
    int min = root.value;
    while (root.left != null) {
      min = root.left.value;
      root = root.left;
    }
    return min;
  }
  
  public int getMax(Node root) {
    int max = root.value;
    while (root.right != null) {
      max = root.right.value;
      root = root.right;
    }
    return max;
  }
  
  public void traverse(Node root) {
    if (root != null) {
      traverse(root.left);
      System.out.println(("VALUE => " + root.value + " | SUB-TREE COUNT => " + root.count));
      traverse(root.right);
    }
  }
  
  void printPostorder(Node node) {
    if (node == null)
      return;
    
    printPostorder(node.left);
    printPostorder(node.right);
    System.out.print(node.value + " ");
  }
  
  void printInorder(Node node) {
    if (node == null)
      return;
    
    printInorder(node.left);
    System.out.print(node.value + " ");
    printInorder(node.right);
  }
  
  void printPreorder(Node node) {
    if (node == null)
      return;
    
    System.out.print(node.value + " ");
    printPreorder(node.left);
    printPreorder(node.right);
  }
  
  /**
   * Getting list of integers
   */
  
  public List<Integer> inorderTraversal(Node root) {
    if (root == null) return list;
    
    inorderTraversal(root.left);
    list.add(root.value);
    inorderTraversal(root.right);
    
    return list;
  }
}

File: src/js/tree/BinarySearchTree.js
/**
 * Binary Search Tree is a node-based binary tree data structure which has the following properties:
 * - The left subtree of a node contains only nodes with keys lesser than the node’s key.
 * - The right subtree of a node contains only nodes with keys greater than the node’s key.
 * - The left and right subtree each must also be a binary search tree.
 *
 */

class Node {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
    this.parent= null;
    this.count = null;
  }
}

class Tree {
  constructor() {
    this.root = null;
  }

  size(node) {
    if (node == null) return 0;
    return node.count;
  }

  /* Insertion into BST*/
  insertUtil(node, value) {
    if (node === null) {
      return new Node(value);
    }
    if (value < node.value) {
      node.left = this.insertUtil(node.left, value);
      node.left.parent = node;
    } else if (value > node.value) {
      node.right = this.insertUtil(node.right, value);
      node.right.parent = node;
    } else {
      node.value = value;
    }
    node.count = this.size(node.left) + this.size(node.right) + 1;
    return node;
  }

  insert(value) {
    this.root = this.insertUtil(this.root, value);
  }

  /*Deletion from BST*/
  deleteUtil(node, value) {
    if (node === null) {
      console.log(`No such node found to delete for value[${value}]`);
      return null;
    };

    if (value < node.value) {
      node.left = this.deleteUtil(node.left, value);
    } else if (value > node.value) {
      node.right = this.deleteUtil(node.right, value);
    } else {
      if (node.left === null) {
        return node.right;
      } else if (node.right === null) {
        return node.left;
      }

      node.value = this.getMin(node.right);
      node.right = this.deleteUtil(node.right, node.value);
    }
    node.count = this.size(node.left) + this.size(node.right) + 1;
    return node;
  }

  delete(value) {
    this.root = this.deleteUtil(this.root, value);
  }

  deleteMinUtil(node) {
    if (node.left === null) return node.right;
    node.left = this.deleteMinUtil(node.left);
    node.count = 1 + this.size(node.left) + this.size(node.right);
    return node;
  }

  deleteMin() {
    this.root = this.deleteMinUtil(this.root);
  }

  deleteMaxUtil(node) {
    if (node.right === null) return node.left;
    node.right = this.deleteMaxUtil(node.right);
    node.count = 1 + this.size(node.right) + this.size(node.left);
    return node;
  }

  deleteMax() {
    this.root = this.deleteMaxUtil(this.root);
  }

  /*Get utils*/
  getUtil(node, value) {
    if (node === null) {
      console.log(`No such node found for value[${value}]`);
      return null;
    } else if (value < node.value) {
      return this.getUtil(node.left, value);
    } else if(value > node.value) {
      return this.getUtil(node.right, value);
    } else {
      return node;
    }
  }

  get(value) {
    return this.getUtil(this.root, value);
  }

  getMinUtil(node) {
    let min = node.value;
    while (node.left !== null) {
      min = node.left.value;
      node = node.left;
    }
    return min;
  }

  getMin() { return this.getMinUtil(this.root) }

  getMaxUtil(node) {
    let max = node.value;
    while (node.right != null) {
      max = node.right.value;
      node = node.right;
    }
    return max;
  }

  getMax() { return this.getMaxUtil(this.root) }
}

module.exports = { Node, Tree };
